Tablecraft Fix and recipe detection

phil235 opened the pull request:
- Fixes notasandwich and flamethrower tablecraft recipes. Fixes #8292. Fixes #8293.
- Add a recipe detection to tablecrafting, if you have some of the things required for the recipe (50% of the amount of different things needed, i.e if less than 3 different things in the recipe then you need 1, otherwise 2+), it will now appear in grey in the tablecrafting window. 
  The tablecrafting window will also lists the requirements for each recipes it shows.
-  Also fixing the flamethrower recipe using 2 rods and the constructed flamethrower not having its weldtool and igniter var set up.
- Also added a fix for the lack of message when trying to turn on an unsecured welding tool, and another fix for the lack of icon update when clicking plasma tank with empty flamethrower (to load it in the flamethrower.

phil235 committed:
Fixes notasandwich and flamethrower tablecraft recipes
Add a recipe detection to tablecrafting, if you have some of the things required for the recipe (50% of the amount of different things needed, i.e if less than 3 different things in the recipe then you need 1, otherwise 2+), it will now appear in grey in the tablecrafting window. The tablecrafting window will also lists the requirements for each recipes it shows.

WJohn commented:
Yay the start of tablecrafting improvements. Though it still needs tons more to be in any way usable.

phil235 commented:
I don't understand Travis' error

RemieRichards commented:
The error is with Travis, not with you.

PKPenguin321 commented:
>if less than 3 different things in the recipe then you need 1, otherwise 2+), it will now appear in grey in the tablecrafting window. 
this is great, table crafting might might might actually get use now

WJohn commented:
I kinda feel this is a weirdly toe-dipping experiment for improving table crafting. I think it'd be really just a good start to make it show all the recipes greyed out at all times (with the materials required shown next to it). I mean you don't see the protolathe hiding its recipes behind a materials barrier any (research barrier sure), then why should tablecrafting?

Jordie0608 commented on code/modules/crafting/recipes.dm:
```diff
@@ -84,10 +84,12 @@
 /datum/table_recipe/flamethrower
 	name = "Flamethrower"
-	result = /obj/item/weapon/flamethrower
+	result = /obj/item/weapon/flamethrower/full
```
Since the /full child of flamethrower is always created with a generic welder, this'll mean when you deconstruct a tablecrafted flamethrower made with an experimental or industrial welder you'll lose that.

phil235 replied:
I fixed it in flamethrower/full/New()

Jordie0608 replied:
Ah my bad, snazzy.

phil235 committed:
Fix for flamethrower deconstruction.

Cheridan commented on code/modules/crafting/table.dm:
```diff
@@ -164,11 +175,46 @@
+						qdel(RE)
+
+				if(R.chem_catalists.len)
+					catalist_text += ", Catalists:"
```
The spelling is Catalyst, unless you're making a pun since this is a list of objects.

phil235 replied:
The typo is in the var name as well so I copied it without thinking, I'll ifx it.

Razharas commented:
Theres actually a proc called on a resulting item after they are crafted that is supposed to handle all kinds of things placed into the resulting item and so on
All atoms have it, its called UpdateParts(if im not mistaken)

phil235 committed:
Fix typo in catalyst.
Fix flamethrower crafting recipe (now only use 1 rod, calls checkparts() to set up the weldingtool and igniter).
Fix lack of message when trying to turn on an unsecured welding tool.
Fix lack of icon update when clicking plasma tank with empty flamethrower.

phil235 commented:
I've changed the fix to use checkparts() like Raz suggested. Also fixing the recipe using 2 rods instead of 1 like the manual construction.

Also added a fix for the lack of message when trying to turn on an unsecured welding tool, and another for the lack of icon update when clicking plasma tank with empty flamethrower (to load it in the flamethrower.

WJohn commented:
Still wish tablecrafting just told you all the recipes but greyed them out if you didn't have all the parts for em. I don't understand why they're being hidden behind a materials barrier, even the protolathe doesn't do that.
